MTM-52191: latest measurement values feature annoucement (#252)
SoftwareAG/c8y-releasenotes: content/release-2024-0/announcements-2024-0.md

##### *[Feature Preview]* Latest measurement values can be stored as part of Device Managed Object
**Important:** The feature is at Public Preview stage means that it's not enabled by default and maybe subject to change in future.

Starting from this release we are introducing the support of automated persistence of measurement values under `c8y_LatestMeasurements` fragment.

##### How to enable it
Using tenant options you should create a category named `measurement.series.latestvalue` with put request to [ tenant options category ](https://cumulocity.com/api/core/#operation/putCategoryOptionResource)
Example: 
```
PUT /tenant/options/measurement.series.latestvalue
{
  "c8y_Humidity.H":"", // to enable single series c8y_Humidity.H
  "c8y_Temperature.*":"", // to able series under fragment c8y_Temperature
  // or "*":"" to enable all 
}
```
Where the key is a filter of measurement series that should be persistent and value must be always empty string, it's left for future use case.

##### How it works

When measurement is created with series that are matching to configuration then device managed object

> will be updated with last series send to platform.

Reviewer: ```suggestion
is updated with the last series sent to the platform.
```

> E.g. when you send 

Reviewer: ```suggestion
Example:

If you send 
```

```
POST /measurement/measurements
{
  "source":"5413"
  "time":"2024-02-01T10:00:00Z"
  "c8y_Temperature":{
     "T": {
        "value": 15,
        "unit":"C"
     }
  }
  "c8y_Speed":{
    "S": {
      "value": 15,
      "unit":"m/s"
    }
  }
}
```

> Then if you consider our example configuration only `c8y_Temperature.T` will be stored as part of device,the `c8y_Speed.S` will be ignored.

Reviewer: ```suggestion
then,  considering the example configuration, only `c8y_Temperature.T` is stored as part of the device, while `c8y_Speed.S` is ignored.
```

> Means measurement will be stored as before, only the state update is skipped.

Reviewer: ```suggestion
This means, that the measurement is stored like before, only the state update is skipped.
```

> To read the latest values on device level you need to use inventory api.

Reviewer: ```suggestion
To read the latest values on device level you must use the Inventory API.
```

> To get single device:

Reviewer: ```suggestion
To get a single device:
```

```
GET /inventory/managedObjects/5413?withLatestValues=true
{
   ...
   "c8y_LatestMeasurements":{
        "c8y_Temperature":{
           "T":{
             "value":15,
             "time":"2024-02-01T10:00:00Z",
             "unit":"C"
           }
        }
   }
}

```

> or to get list of devices matching expected criteria 

Reviewer: ```suggestion
To get a list of devices matching the expected criteria, 
```

> e.g. get all devices which have reported temperature higher than 10 degrees :

Reviewer: ```suggestion
for example, get all devices which have a reported temperature higher than 10 degrees:
```

```
GET /inventory/managedObjects?withLatestValues=true&query=$filter=c8y_LatestMeasurements.c8y_Temperature.T.value+gt+10
{
  managedObjects: [
    {
        ...
        "c8y_LatestMeasurements":{
            "c8y_Temperature":{
                "T":{
                    "value":15,
                    "time":"2024-02-01T10:00:00Z",
                    "unit":"C"
                }
            }
        }
    }
  ]
}

```
##### Implications

> The feature is introducing addition operation upon measurement creation.

Reviewer: ```suggestion
This feature introduces an additional operation upon measurement creation.
```

> That will result in performance degradation, which depends on number of series that needs to be 

Reviewer: ```suggestion
This results in performance degradation, depending on the number of series to be
```

> stored with each and goes from 5% for single series in each measurement to 

Reviewer: ```suggestion
stored in each measurement, reaching from 5% for single series in each measurement to 
```

more than 20% in case of 50 series per measurement.

#### Limitations

##### Security

> The latest measurements values are part of managed object, and they follow the managed object inventory role permissions.

Reviewer: ```suggestion
The latest measurement values are part of the managed object and they follow the managed object inventory role permissions.
```

> It's not respecting the inventory roles for measurements. 

Reviewer: Where does "it" refer to?

Author: It's a continuation of previous sentence, I'm still unsure how to put words in correct way:  ` they are not respecting the inventory roles for measurements but instead it follows the managed object inventory role permissions` 
Would it be better ?

Reviewer 2: added a suggestion as a separate comment

##### Data model

The latest measurements are not store the measurement type this information is lost
and can be found only using original measurement document.

#### Last value

> The value stored at device managed object is the last value send to platform. 

Reviewer: ```suggestion
The value stored in the device managed object is the last value sent to the platform. 
```

```
GET /inventory/managedObjects/5413?withLatestValues=true
{
   ...
   "c8y_LatestMeasurements":{
        "c8y_Temperature":{
           "T":{
             "value":15,
             "time":"2024-02-01T10:00:00Z",
             "unit":"C"
           }
        }
   }
}

```
To get a list of devices matching the expected criteria, 
for example, get all devices which have a reported temperature higher than 10 degrees:

```
GET /inventory/managedObjects?withLatestValues=true&query=$filter=c8y_LatestMeasurements.c8y_Temperature.T.value+gt+10
{
  managedObjects: [
    {
        ...
        "c8y_LatestMeasurements":{
            "c8y_Temperature":{
                "T":{
                    "value":15,
                    "time":"2024-02-01T10:00:00Z",
                    "unit":"C"
                }
            }
        }
    }
  ]
}

```
##### Implications

This feature introduces an additional operation upon measurement creation.
This results in performance degradation, depending on the number of series to be
stored in each measurement, reaching from 5% for single series in each measurement to 
more than 20% in case of 50 series per measurement.

#### Limitations

##### Security

> The latest measurement values are part of the managed object and they follow the managed object inventory role permissions.
> It's not respecting the inventory roles for measurements. 

Reviewer 2: ```suggestion
The latest measurement values are part of the managed object and they follow the managed object inventory role permissions instead of respecting the inventory roles for measurements. 
```

stored with each and goes from 5% for single series in each measurement to 
more than 20% in case of 50 series per measurement.

#### Limitations

##### Security

The latest measurements values are part of managed object, and they follow the managed object inventory role permissions.
It's not respecting the inventory roles for measurements. 

##### Data model

> The latest measurements are not store the measurement type this information is lost

Reviewer 3: ```suggestion
The latest measurements do not store the measurement type. This information
```

> and can be found only using original measurement document.

Reviewer 3: ```suggestion
can be obtained using Measurements API.
```

Reviewer 2: edited the suggestion

#### Last value

The value stored at device managed object is the last value send to platform. 

> Means that if the order of measurement delivery to platform is not correct 

Reviewer 3: ```suggestion
When the order of measurement delivery to the platform is different than measurement creation time
```

Reviewer 2: edited the suggestion
